core/integration/cache_test.go

```diff
@@ -107,3 +107,235 @@ func (CacheSuite) TestLocalImportCacheReuse(ctx context.Context, t *testctx.T) {
+		WithNewFile("main.go", fooTmpl)
+
+	fooID, err := ctr.
+		WithWorkdir("/work/foo").
+		With(daggerExec("call", "get-cache-volume-id")).
+		Stdout(ctx)
+	require.NoError(t, err)
+
+	barID, err := ctr.
+		WithWorkdir("/work/bar").
+		With(daggerExec("call", "get-cache-volume-id")).
+		Stdout(ctx)
+
+	require.NoError(t, err)
+	require.NotEqual(t, fooID, barID)
+}
+
+func (CacheSuite) TestCacheIdSameAcrossSession(ctx context.Context, t *testctx.T) {
+	session1 := connect(ctx, t)
+
+	fooTmpl := `package main
+	import (
+		"context"
+	)
+
+	type Foo struct {}
+	func (f *Foo) GetCacheVolumeId(ctx context.Context) (string, error) {
+		id, err := dag.CacheVolume("volume-name").ID(ctx)
+		return string(id), err
+	}
+	`
+
+	ctr1 := session1.Container().
+		From(golangImage).
+		WithMountedFile(testCLIBinPath, daggerCliFile(t, session1)).
+		WithWorkdir("/work/foo").
+		With(daggerExec("init", "--name=foo", "--source=.", "--sdk=go")).
+		WithNewFile("main.go", fooTmpl)
+
+	fooID, err := ctr1.
+		WithWorkdir("/work/foo").
+		With(daggerExec("call", "get-cache-volume-id")).
+		Stdout(ctx)
+	require.NoError(t, err)
+
+	session2 := connect(ctx, t)
+	ctr2 := session2.Container().
+		From(golangImage).
+		WithMountedFile(testCLIBinPath, daggerCliFile(t, session2)).
+		WithWorkdir("/work/foo").
+		With(daggerExec("init", "--name=foo", "--source=.", "--sdk=go")).
+		WithNewFile("main.go", fooTmpl)
+
+	fooID2, err := ctr2.
+		WithWorkdir("/work/foo").
+		With(daggerExec("call", "get-cache-volume-id")).
+		Stdout(ctx)
+
+	require.NoError(t, err)
+	require.Equal(t, fooID, fooID2)
+}
+
+func (CacheSuite) TestCacheVolumePassedAcrossModules(ctx context.Context, t *testctx.T) {
+	session := connect(ctx, t)
+
+	fooTmpl := `package main
+import (
+	"context"
+	"dagger/foo/internal/dagger"
+	"fmt"
+)
+
+type Foo struct {}
+	
+func (f *Foo) Populate(ctx context.Context, input string) (*dagger.Container, error) {
+	return dag.Container().
+		From("alpine").
+		WithMountedCache("/tmp-cache-mount", dag.CacheVolume("cache-name")).
+		WithExec([]string{"sh", "-c", fmt.Sprintf("echo '%s' > /tmp-cache-mount/input.txt", input)}).
+		Sync(ctx)
+}
+
+func (f *Foo) Fetch(ctx context.Context) (string, error) {
+	cache := dag.CacheVolume("cache-name")
+	return dag.Bar().Fetch(ctx, cache)
+}
+`
+
+	barTmpl := `package main
+	import (
+		"context"
+		"dagger/bar/internal/dagger"
+	)
+
+	type Bar struct {}
+func (f *Bar) Fetch(ctx context.Context, vol *dagger.CacheVolume) (string, error) {
+	return dag.Container().
+		From("alpine").
+		WithMountedCache("/tmp-cache-mount", vol).
+		WithExec([]string{"sh", "-c", "cat /tmp-cache-mount/input.txt"}).
+		Stdout(ctx)
+}
+
+func (f *Bar) FetchDirect(ctx context.Context) (string, error) {
+vol := dag.CacheVolume("cache-name")
+	return dag.Container().
+		From("alpine").
+		WithMountedCache("/tmp-cache-mount", vol).
+		WithExec([]string{"sh", "-c", "cat /tmp-cache-mount/input.txt"}).
```

Comment:
Hi @sipsma, I think this scenario should work because we are passing along the cache volume to `Bar` from `Foo`. do you still think that this testcase should have failed? if so, did I misunderstood the testcase you mentioned in other comment:

> We'll want an additional test where a function in one module passes a cache volume to a function in a different module and make sure that it works as expected (cache volume should be tied to the one who created it, not used it)

OR It maybe that this comment is showing up on a different line number now that I've pushed some more changes.

Reply:
Ah no, not sure if I misread it when I left the comment or if something changed, but looking again now this seems like it's testing+working as expected 👍

Reply:
just to avoid any confusion around caching, I changed the volume mount to a diff path.  👍